### PR TITLE
Promisify the internal imdb interface

### PR DESCRIPTION
**Pull request by FenrirUnbound**
The `[datastore-base](https://github.com/screwdriver-cd/datastore-base)` has updated to version 2, which implements a Promise-based interface. Because the DynamoDB counterpart is already implementing this, screwdriver-cd/screwdriver#205 is dropping this `datastore-imdb` module entirely.

This PR is to update the `datastore-imdb` module to adhere to the new Promise-based interface.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -92,18 +92,20 @@ class MyDatastore extends Datastore {
...
-        if (!data) {
-            return callback(null, null);
-        }
-        data = hoek.merge(data, config.params.data, false);
-        imdb.set(config.table, config.params.id, data);
+            const result = hoek.merge(data, config.params.data, false);
 
-        return callback(null, data);
+            return imdb.set(config.table, config.params.id, result)
+                .then(() => data);
```
Comment (tkyi) [inline index.js]: should it be `result` that is returned, not `data`?

Comment (FenrirUnbound) [inline index.js]: It should actually be the value returned from `imdb.set`. Good catch 👍

Comment (FenrirUnbound) [inline index.js]: Your concern has been properly address 😺

File: index.js [hunk 2]
```diff
@@ -92,18 +92,20 @@ class MyDatastore extends Datastore {
...
-        if (!data) {
-            return callback(null, null);
-        }
-        data = hoek.merge(data, config.params.data, false);
-        imdb.set(config.table, config.params.id, data);
+            const updatedData = hoek.merge(data, config.params.data, false);
 
-        return callback(null, data);
+            return imdb.set(config.table, config.params.id, updatedData)
+                .then((result) => result);
```
Comment (d2lam) [inline index.js]: dont think you need the `.then((result) => result)` here

File: lib/imdb.js [hunk 3]
```diff
@@ -9,28 +9,29 @@ let tables = {};
  * @method getValue
  * @param  {String} tableName Name of the datastore table
  * @param  {String} key       Key of the row to obtain the value from
- * @return {Object}           The value in the datastore
+ * @return {Promise}          A promise that resolves into the value in the datastore
  */
 function getValue(tableName, key) {
     if (!tables[tableName]) {
-        return null;
+        return Promise.resolve(null);
```
Comment (d2lam) [inline lib/imdb.js]: Shouldn't this be a reject to be consistent with dynamo?
https://github.com/screwdriver-cd/datastore-dynamodb/blob/master/index.js#L61

File: README.md [hunk 4]
```diff
@@ -92,19 +91,15 @@ imdb.get({
     params: {
         id: 'apple'
     }
-}, (err, data) => {
-    console.log(data); // { color: 'red', type: 'fruit' }
-});
+}).then(console.log);  // { color: 'red', type: 'fruit' }
```
Comment (FenrirUnbound) [inline README.md]: `console.log` is a function. It passes `data` to it as a parameter

File: index.js (hunk 2 above)
Comment (FenrirUnbound) [inline index.js]: 👍

File: lib/imdb.js (hunk 3 above)
Comment (FenrirUnbound) [inline lib/imdb.js]: It's consistent enough with `datastore-base`: https://github.com/screwdriver-cd/datastore-base#get

If the table does not exist, the value does not exist. Therefore, it should return `null`. If you believe this should be an error reflected by all modules that extend it, we should update the base model

Comment (d2lam) [inline lib/imdb.js]: I think it's more intuitive to return error for invalid table name. Like in SQL if you try to query from an invalid table, it also returns error.

File: README.md (hunk 4 above)
Comment (d2lam) [inline README.md]: Oh that's cool. I didn't know that 👍

File: lib/imdb.js (hunk 3 above)
Comment (FenrirUnbound) [inline lib/imdb.js]: It's been added

File: index.js [hunk 5]
```diff
@@ -115,18 +116,18 @@ class MyDatastore extends Datastore {
...
-        if (!data) {
-            const err = new Error(`Invalid table name "${config.table}"`);
-
-            return callback(err);
-        }
+    _scan(config) {
+        return imdb.scan(config.table, config.params, config.paginate)
+        .then((data) => {
+            if (!data) {
+                throw new Error(`Invalid table name "${config.table}"`);
```
Comment (stjohnjohnson) [inline index.js]: imdb.scan already rejects, this shouldn't reach here, see line 16 of imdb

Comment (FenrirUnbound) [inline index.js]: Resolved

**Comment by FenrirUnbound** [comment]
@d2lam @tkyi @stjohnjohnson your follow-up review is needed

**Comment by d2lam** [comment]
LGTM 👍

File: index.js [hunk 6]
```diff
@@ -115,18 +116,12 @@ class MyDatastore extends Datastore {
...
-        if (!data) {
-            const err = new Error(`Invalid table name "${config.table}"`);
-
-            return callback(err);
-        }
-
-        return callback(null, data);
+    _scan(config) {
+        return imdb.scan(config.table, config.params, config.paginate)
+        .then((data) => data || []);
```
Comment (stjohnjohnson) [inline index.js]: Just curious, when would data ever be null anymore?

Comment (FenrirUnbound) [inline index.js]: Probably never. It's a defensive check.

**Pull request merged**